packages/app/src/cli/services/dev/extension/websocket/handlers.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  EventType,
+  IncomingDispatchMessage,
+  OutgoingDispatchMessage,
+  OutgoingMessage,
+  SetupWebSocketConnectionOptions,
+} from './models.js'
+import {RawData, WebSocket, WebSocketServer} from 'ws'
+import {http, output} from '@shopify/cli-kit'
+import {Duplex} from 'stream'
+
+export function websocketUpgradeHandler(
+  wss: WebSocketServer,
+  options: SetupWebSocketConnectionOptions,
+): (req: http.IncomingMessage, socket: Duplex, head: Buffer) => void {
+  return (request, socket, head) => {
+    if (request.url !== '/extensions') {
+      return
+    }
+    output.debug(`Upgrading HTTP request to a websocket connection`)
+    wss.handleUpgrade(request, socket, head, getConnectionDoneHandler(wss, options))
+  }
+}
+
+export function getConnectionDoneHandler(wss: WebSocketServer, options: SetupWebSocketConnectionOptions) {
+  return (ws: WebSocket) => {
+    output.debug(`Websocket connection successfully established`)
+    const connectedPayload = {
+      event: 'connected',
+      data: options.payloadStore.getConnectedPayload(),
+      version: '3',
+    }
+    output.debug(output.content`Sending connected payload: ${output.token.json(connectedPayload)}`)
+    ws.send(JSON.stringify(connectedPayload))
+    ws.on('message', getOnMessageHandler(wss, options))
+  }
+}
+
+export function getOnMessageHandler(wss: WebSocketServer, options: SetupWebSocketConnectionOptions) {
+  return (data: RawData) => {
+    const jsonData = JSON.parse(data.toString())
+    const {event: eventType, data: eventData} = jsonData
+
+    output.debug(output.content`Received websocket message with event type ${eventType} and data:
+${output.token.json(eventData)}
+          `)
+
+    if (eventType === 'update') {
+      /**
+       * App updates must take precedence over extensions. Otherwise the websocket server
+       * will send an update to the client with missing app data and will cause the loading
+       * of extensions to fail.
+       */
+      if (eventData.app) {
```

Comment:
This is missing a check to verify that the app.apiKey matches the one in memory - otherwise the update should be ignored

Reply:
Done